src/lib/fcitx/inputmethodengine.cpp

```diff
@@ -28,18 +27,43 @@ std::string InputMethodEngine::subModeLabel(const InputMethodEntry &entry,
 
 void InputMethodEngine::virtualKeyboardEvent(
     const InputMethodEntry &entry, VirtualKeyboardEvent &virtualKeyboardEvent) {
-    if (auto *this4 = dynamic_cast<InputMethodEngineV4 *>(this)) {
-        this4->virtualKeyboardEventImpl(entry, virtualKeyboardEvent);
-    } else if (auto virtualKeyEvent = virtualKeyboardEvent.toKeyEvent()) {
-        keyEvent(entry, *virtualKeyEvent);
-        // TODO: revisit the default action.
-        if (virtualKeyEvent->accepted()) {
-            virtualKeyboardEvent.accept();
-        } else if (!virtualKeyboardEvent.text().empty()) {
-            virtualKeyboardEvent.inputContext()->commitString(
-                virtualKeyboardEvent.text());
-        }
+    auto *this4 = dynamic_cast<InputMethodEngineV4 *>(this);
+    if (this4 == nullptr) {
+        FCITX_WARN() << "InputMethodEngine::virtualKeyboardEvent is called."
+                     << "But there is no valid InputMethodEngineV4 instance."
+                     << "So it returns early.";
+        return;
     }
+
+    this4->virtualKeyboardEventImpl(entry, virtualKeyboardEvent);
+}
+
+FunctionMode InputMethodEngine::getFunctionMode() const {
```

Comment:
I don't quite understand why function mode needs a setter function. shouldn't it be a static property?

Reply:
It seems that the mode is a instance property, then shouldn't engine just read this property?

Reply:
The input method engine can support just full function mode or support both full function mode and limited funciton. So we should tell the input method engine which function mode should be effective. Or else, the input method engine should use full function mode when keyEvent function is caled and limited function mode when virtualKeyboardEvent function is called.